ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+* `Ractor`s will be used for state isolation by default assuming decent performance but easy to disable and mostly
+  hidden from users.
+
+These points are explained below.
+
+## Workflow Definition/Usage
+
+Examples below are taken from
+[this Python sample](https://github.com/temporalio/samples-python/tree/main/message_passing/introduction), so not too
+many comments about general non-Ruby behavior are inline.
+
+Example definition:
+
+```ruby
+class CallGreetingServiceActivity < Temporalio::Activity::Definition
+  GREETINGS = {
+    arabic: 'مرحبا بالعالم',
+    chinese: '你好，世界',
+    english: 'Hello, world',
+    french: 'Bonjour, monde',
+    hindi: 'नमस्ते दुनिया',
+    portuguese: 'Olá mundo',
+    spanish: 'Hola mundo'
+  }
+
+  def execute(language)
+    # Simulate a network call
+    sleep(0.2)
+    GREETINGS[language.to_sym]
+  end
+end
+
+class GreetingWorkflow < Temporalio::Workflow::Definition
+  workflow_query_attr :language
+
+  def initialize(name)
+    @greetings = {
+      chinese: '你好，世界',
+      english: 'Hello, world'
+    }
+    @language = :english
+    @mutex = Temporalio::Workflow::Mutex.new
+  end
+
+  def execute
+    Temporalio::Workflow.wait_condition do
+      @approved_for_release && Temporalio::Workflow.all_handlers_finished?
+    end
+    @greetings[@language]
+  end
+
+  workflow_query
+  def languages(input)
+    return CallGreetingServiceActivity::GREETINGS.keys.sort if input[:include_unsupported]
+    @greetings.keys.sort
+  end
+
+  workflow_signal
+  def approve(input)
+    @approved_for_release = true
+    @approver_name = input[:name]
+  end
+
+  workflow_update
+  def set_language(language)
+    previous, @language = @language, language.to_sym
+    previous
+  end
+
+  workflow_update_validator(:set_language)
+  def validate_set_language(language)
+    raise "#{language} not supported" unless @greetings.include?(language.to_sym)
+  end
+
+  workflow_update
+  def set_language_using_activity(language)
+    language = language.to_sym
+    unless @greetings.include?(language)
+      @mutex.synchronize do
+        greeting = Temporalio::Workflow.execute_activity(
+          CallGreetingServiceActivity,
+          language,
+          start_to_close_timeout: 10
+        )
+        unless greeting
+          raise Temporalio::Error::ApplicationError, "Greeting service does not support #{language}"
+        end
+        @greetings[language] = greeting
+      end
+    end
+    previous, @language = @language, language
+    previous
+  end
+end
+```
+
+Example worker:
+
+```ruby
+client = Temporalio::Client.connect('localhost:7233', 'my-namespace')
+worker = Temporalio::Worker.new(
+  client:,
+  task_queue: 'my-task-queue',
+  activities: [GreetingActivity],
+  workflows: [GreetingWorkflow]
+)
+worker.run(shutdown_signals: ['SIGINT'])
+```
+
+Example client usage:
+
+```ruby
+client = Temporalio::Client.connect('localhost:7233', 'my-namespace')
+
+# Start workflow
+wf_handle = client.start_workflow(
```

Comment:
re: subclassing `Workflow::Definition`, you mentioned that as a controversial decision.  Is the reason for that so that clients can refer to just the workflow here, without pulling in the definition which might be in a separate module?  If so, that seems like an okay tradeoff to me.
IIRC, coinbase/temporal ruby used the _instance_ to be the implementation specific stuff whereas the _class_ was used by the client.  I don't have a strong idea of how good that was, since Stripe chose not to isolate client concerns from definition concerns, so I don't know if we could have done better given coinbase SDK.  This was a usability win but ended up biting us after a few years after we got religion on packaging modules.